File: examples/Haskell/asyncsrv.hs
-- |
-- Asynchronous client-to-server (DEALER to ROUTER) p.111
-- Compile with -threaded

module Main where

import System.ZMQ4.Monadic
import ZHelpers (setRandomIdentity)
import Control.Concurrent (threadDelay)
import Data.ByteString.Char8 (pack, unpack)
import Control.Monad (forever, forM_, replicateM_)
import System.Random (randomRIO)
import Text.Printf

clientTask :: String -> ZMQ z ()
clientTask ident = do
    client <- socket Dealer
    setRandomIdentity client
    connect client "tcp://localhost:5570"
    forM_ [1..] $ \i -> do -- (long enough) forever
        -- tick one per second, pulling in arriving messages
        forM_ [0..100] $ \_ -> 
            poll 10 -- timeout of 10 ms
                [Sock client [In] -- wait for incoming event 
                $ Just $ -- if it happens do
                    \_ -> receive client >>= liftIO . printf "Client %s has received back from worker its msg \"%s\"\n" ident . unpack ] 
              
        send client [] (pack $ unwords ["Client", ident, "sends request", show i])

            
serverTask :: ZMQ z ()
serverTask =  do
    frontend <- socket Router
    bind frontend "tcp://*:5570"
    backend <- socket Dealer
    bind backend "inproc://backend"
    
    replicateM_ 5 $ async serverWorker

    proxy frontend backend Nothing


serverWorker :: ZMQ z ()
serverWorker = do
    worker <- socket Dealer
    connect worker "inproc://backend"
    liftIO $ putStrLn "Worker Started"        
    forever $ -- receive both ident and msg and send back the msg to the ident client.
        receive worker >>= \ident -> receive worker >>= \msg -> sendback worker msg ident

    where
         -- send back to client 0 to 4 times max
        sendback worker msg ident  = do
            resentNb <- liftIO $ randomRIO (0, 4)
            timeoutMsec <- liftIO $ randomRIO (1, 1000) 
            forM_ [0::Int ..resentNb] $ \_ -> do
                liftIO $ threadDelay $ timeoutMsec * 1000
                send worker [SendMore] ident
                send worker [] msg

main :: IO ()
main = 
    runZMQ $ do
        async $ clientTask "A"
        async $ clientTask "B"
        async $ clientTask "C"
        async serverTask
        
        liftIO $ threadDelay $ 5 * 1000 * 1000

File: examples/Ruby/asyncsrv.rb
# Asynchronous client-to-server (DEALER to ROUTER)

require 'rubygems'
require 'ffi-rzmq'

def client
  context = ZMQ::Context.new
  client = context.socket ZMQ::DEALER
  client.identity = "%04X-%04X" % [rand(0x10000), rand(0x10000)]
  client.connect "ipc://frontend.ipc"

  poller = ZMQ::Poller.new
  poller.register_readable(client)

  request_number = 0

  loop do

    100.times do |tick|
      if poller.poll(10) == 1
        client.recv_strings message = []
        puts "#{client.identity}: #{message.last}"
      end
    end

    client.send_string "Req ##{request_number += 1}"
  end

  client.close
  context.destroy
end

def worker(context)
  worker = context.socket ZMQ::DEALER
  worker.connect "inproc://backend"

  loop do
    worker.recv_strings message = []

    rand(0..4).times do
      sleep rand
      worker.send_strings message
    end
  end

  worker.close
end

def server
  context = ZMQ::Context.new
  frontend = context.socket ZMQ::ROUTER
  backend = context.socket ZMQ::DEALER

  frontend.bind "ipc://frontend.ipc"
  backend.bind "inproc://backend"

  poller = ZMQ::Poller.new
  poller.register_readable frontend
  poller.register_readable backend

  5.times { Thread.new { worker context } }

  ZMQ::Device.create ZMQ::QUEUE, frontend, backend

end

3.times { Thread.new { client } }

server
